server/chat-plugins/the-studio.ts

```diff
@@ -0,0 +1,613 @@
+		accountName = accountName.trim();
+		const sanitizedName = accountName.replace(/[^-_a-zA-Z0-9]+/g, '');
+		if (!(!accountName.includes(' ') && accountName === sanitizedName && /^[a-zA-Z]/.test(sanitizedName) &&
+			sanitizedName.length > 1 && sanitizedName.length < 16)) {
+			throw new Chat.ErrorMessage(`The provided account name (${sanitizedName}) is invalid. Valid last.fm usernames are between 2-15 characters, start with a letter, and only contain letters, numbers, hyphens, and underscores.`);
+		}
+		return true;
+	}
+
+	getAccountName(username: string) {
+		if (lastfm[toID(username)]) return lastfm[toID(username)];
+		return username.trim().replace(/ /g, '_').replace(/[^-_a-zA-Z0-9]/g, '');
+	}
+
+	async tryGetTrackData(track: string, artist?: string) {
+		this.checkHasKey();
+		const query: {[k: string]: any} = {
+			method: 'track.search', limit: 1, api_key: Config.lastfmkey, track, format: 'json',
+		};
+		if (artist) query.artist = artist;
+		const raw = await Net(API_ROOT).get({query});
+		const req = JSON.parse(raw);
+		let buf = ``;
+		if (req.results?.trackmatches?.track?.length) {
+			buf += `<table><tr><td style="padding-right:5px">`;
+			const obj = req.results.trackmatches.track[0];
+			const trackName = obj.name || "Untitled";
+			const artistName = obj.artist || "Unknown Artist";
+			const searchName = `${artistName} - ${trackName}`;
+			if (obj.image?.length) {
+				const img = obj.image;
+				const imageIndex = img.length >= 3 ? 2 : img.length - 1;
+				if (img[imageIndex]['#text'] && !DEFAULT_IMAGES.includes(img[imageIndex]['#text'])) {
+					buf += `<img src="${img[imageIndex]['#text']}" width="75" height="75" />`;
+				}
+			}
+			buf += `</td><td>`;
+			const artistUrl = obj.url.split('_/')[0];
+			buf += `<strong><a href="${artistUrl}">${artistName}</a> - <a href="${obj.url}">${trackName}</a></strong><br />`;
+			const videoIDs = await YouTube.searchVideo(searchName, 1);
+			if (!videoIDs?.length) {
+				buf += searchName;
+			} else {
+				buf += `<a href="https://youtu.be/${videoIDs[0]}">YouTube link</a>`;
+			}
+			buf += `</td></tr></table>${this.getScrobbleBadge()}`;
+		}
+		if (req.error) {
+			throw new Chat.ErrorMessage(`${req.message}.`);
+		}
+		if (!buf) {
+			throw new Chat.ErrorMessage(`No results for '${artist ? `${artist} - ` : ``}${track}' found. Check spelling?`);
+		}
+		return buf;
+	}
+	checkHasKey() {
+		if (!Config.lastfmkey) {
+			throw new Chat.ErrorMessage(`This server does not support last.fm commands. If you're the owner, you can enable them by setting up Config.lastfmkey.`);
+		}
+	}
+	getScrobbleBadge() {
+		return `<div style="float:right;color:#888;font-size:8pt">[powered by AudioScrobbler]</div><div style="clear:both"></div>`;
+	}
+}
+
+class RecommendationsInterface {
+	getRandomRecommendation() {
+		const recs = recommendations.saved;
+		return recs[Math.floor(Math.random() * recs.length)];
+	}
+
+	checkRoom(room: Room | null) {
+		if (!room || room.roomid !== 'thestudio') {
```

Comment:
```suggestion
		if (room?.roomid !== 'thestudio') {
```
Also I don't necessarily think this should be in a class; it seems more like a helper for the chat commands that could go at the top level rather than something for recommendations specifically.

Reply:
Why shouldn't it be in the Recommendation class? Recommendations are The Studio-specific, so it makes perfect sense in my opinion to put the function to check if the room that the class is used for IN the class.

Reply:
It's also used in other chat plugins similarly, so those would need changed too.

Reply:
I'm fine with it either way; there are chat plugins that put these things at the top level too. that was more of a personal nitpick, i should've made that more clear, mb

Reply:
This should probably just be an argument for `requireRoom` to require a specific roomid.

Actually doesn't `requireRoom` already have that feature? Why does it need to be duplicated?